File: triage_cli/providers/unleash.py
"""Unleash LLM provider — calls e-api.unleash.so /chats."""
from __future__ import annotations

import json
import logging
import os
from typing import Any

import httpx

logger = logging.getLogger(__name__)

DEFAULT_UNLEASH_BASE_URL = "https://e-api.unleash.so"
LLM_TIMEOUT_SECONDS = 90.0


class UnleashProvider:
    name = "unleash"

    async def complete(self, *, prompt: str, system_prompt: str, model: str) -> str:
        """Call Unleash /chats and return concatenated assistant text parts."""
        endpoint = f"{_base_url('UNLEASH_BASE_URL', DEFAULT_UNLEASH_BASE_URL)}/chats"
        headers = _unleash_headers()
        payload = {
            "assistantId": _required_env("UNLEASH_ASSISTANT_ID", provider=self.name),
            "stream": False,
            "messages": [
                {"role": "System", "text": system_prompt},
                {"role": "User", "text": prompt},
            ],
        }

        data, response_headers = await _post_json(endpoint, headers=headers, payload=payload)
        text = _unleash_text_from_response(data)
        if not text:
            request_id = _request_id_from_payload(data) or response_headers.get("RequestId")
            suffix = f" RequestId: {request_id}." if request_id else ""
            raise RuntimeError(
                "Unleash API response did not include any assistant text parts." + suffix
            )
        return text


def _base_url(env_name: str, default: str) -> str:
    return (os.getenv(env_name) or default).strip().rstrip("/")


def _required_env(name: str, *, provider: str) -> str:
    value = os.getenv(name, "").strip()
    if not value:
        raise RuntimeError(f"{name} must be set when LLM_PROVIDER={provider}.")
    return value


def _unleash_headers() -> dict[str, str]:
    api_key = _required_env("UNLEASH_API_KEY", provider="unleash")
    lowered = api_key.lower()
    if lowered.startswith("bearer "):
        authorization = api_key
    elif lowered.startswith("bearer:"):
        authorization = f"Bearer {api_key.split(':', 1)[1].strip()}"
    else:
        authorization = f"Bearer {api_key}"

    headers = {
        "Authorization": authorization,
        "Accept": "application/json",
        "Content-Type": "application/json",
    }
    account = os.getenv("UNLEASH_ACCOUNT", "").strip()
    if account:
        headers["unleash-account"] = account
    return headers


def _unleash_text_from_response(data: Any) -> str:
    """Extract Text parts from a documented Unleash chat response."""
    if isinstance(data, list):
        return "".join(_unleash_text_from_response(item) for item in data)
    if not isinstance(data, dict):
        return ""

    message = data.get("message")
    if not isinstance(message, dict):
        return ""
    parts = message.get("parts")
    if not isinstance(parts, list):
        return ""

    chunks: list[str] = []
    for part in parts:
        if not isinstance(part, dict) or part.get("type") != "Text":
            continue
        text = part.get("text")
        if isinstance(text, str):
            chunks.append(text)
    return "".join(chunks)


def _request_id_from_payload(data: Any) -> str | None:
    if isinstance(data, dict):
        value = data.get("requestId")
        return value if isinstance(value, str) and value else None
    return None


async def _post_json(
    endpoint: str,
    *,
    headers: dict[str, str],
    payload: dict[str, Any],
) -> tuple[Any, dict[str, str]]:
    try:
        async with httpx.AsyncClient(timeout=LLM_TIMEOUT_SECONDS) as client:
            response = await client.post(endpoint, headers=headers, json=payload)
    except httpx.RequestError as e:
        raise RuntimeError(f"LLM provider API call failed: {e}") from e

    if response.status_code >= 400:
        raise RuntimeError(_provider_error_message(response))

    try:
        data = response.json()
    except json.JSONDecodeError as e:
        raise RuntimeError("LLM provider API response was not valid JSON.") from e
    return data, dict(response.headers)


def _provider_error_message(response: httpx.Response) -> str:
    detail = response.text.strip()
    request_id = response.headers.get("RequestId") or response.headers.get("x-request-id")
    try:
        payload = response.json()
    except json.JSONDecodeError:
        payload = None
    if isinstance(payload, dict):
        detail_value = payload.get("message") or payload.get("error") or payload.get("detail")
        if isinstance(detail_value, dict):
            detail = str(detail_value.get("message") or detail_value)
        elif isinstance(detail_value, str):
            detail = detail_value
        request_id = request_id or _request_id_from_payload(payload)
    suffix = f" RequestId: {request_id}." if request_id else ""
    detail_suffix = f": {detail}" if detail else ""
    return f"LLM provider API call failed with HTTP {response.status_code}{detail_suffix}.{suffix}"

File: triage-cli-rs/src/providers/unleash.rs
//! Unleash LLM provider — POST to `/chats`. Mirrors Python `providers.unleash`.

use std::env;
use std::future::Future;
use std::pin::Pin;
use std::time::Duration;

use reqwest::Client;
use serde_json::{json, Value};

use super::{base_url, required_env, CompletionResult, LlmProvider, ProviderError};

const DEFAULT_BASE_URL: &str = "https://e-api.unleash.so";
const LLM_TIMEOUT_SECS: u64 = 90;

pub struct UnleashProvider;

impl LlmProvider for UnleashProvider {
    fn name(&self) -> &'static str {
        "unleash"
    }

    fn complete<'a>(
        &'a self,
        prompt: &'a str,
        system_prompt: &'a str,
        _model: &'a str,
    ) -> Pin<Box<dyn Future<Output = Result<CompletionResult, ProviderError>> + Send + 'a>> {
        Box::pin(async move {
            let endpoint = format!("{}/chats", base_url("UNLEASH_BASE_URL", DEFAULT_BASE_URL));
            let headers = unleash_headers()?;
            let assistant_id = required_env("UNLEASH_ASSISTANT_ID", "unleash")?;
            let payload = json!({
                "assistantId": assistant_id,
                "stream": false,
                "messages": [
                    {"role": "System", "text": system_prompt},
                    {"role": "User", "text": prompt},
                ],
            });
            let (data, response_headers) = post_json(&endpoint, headers, payload).await?;
            let text = unleash_text_from_response(&data);
            if text.is_empty() {
                let rid = request_id_from_payload(&data)
                    .or_else(|| response_headers.get("requestid").cloned())
                    .map(|r| format!(" RequestId: {r}."))
                    .unwrap_or_default();
                return Err(ProviderError::NoText("Unleash", rid));
            }
            let (tokens_in, tokens_out) = unleash_token_counts(&data);
            Ok(CompletionResult {
                text,
                tokens_in,
                tokens_out,
            })
        })
    }
}

fn unleash_headers() -> Result<Vec<(String, String)>, ProviderError> {
    let api_key = required_env("UNLEASH_API_KEY", "unleash")?;
    let lowered = api_key.to_ascii_lowercase();
    let authorization = if lowered.starts_with("bearer ") {
        api_key.clone()
    } else if lowered.starts_with("bearer:") {
        let after = api_key.split_once(':').map(|x| x.1).unwrap_or("").trim();
        format!("Bearer {after}")
    } else {
        format!("Bearer {api_key}")
    };
    let mut headers = vec![
        ("Authorization".into(), authorization),
        ("Accept".into(), "application/json".into()),
        ("Content-Type".into(), "application/json".into()),
    ];
    let account = env::var("UNLEASH_ACCOUNT").unwrap_or_default();
    let account = account.trim();
    if !account.is_empty() {
        headers.push(("unleash-account".into(), account.to_string()));
    }
    Ok(headers)
}

fn unleash_text_from_response(data: &Value) -> String {
    match data {
        Value::Array(items) => items.iter().map(unleash_text_from_response).collect(),
        Value::Object(_) => {
            let message = data.get("message").cloned().unwrap_or(Value::Null);
            let parts = message
                .get("parts")
                .and_then(Value::as_array)
                .cloned()
                .unwrap_or_default();
            let mut chunks = Vec::with_capacity(parts.len());
            for part in parts {
                if part.get("type").and_then(Value::as_str) == Some("Text") {
                    if let Some(s) = part.get("text").and_then(Value::as_str) {
                        chunks.push(s.to_string());
                    }
                }
            }
            chunks.join("")
        }
        _ => String::new(),
    }
}

/// Extract token usage counts from the Unleash response when present.
/// Tries common field paths used by LLM gateway APIs; returns `(None, None)`
/// if the fields are absent (non-fatal — token counts are best-effort).
fn unleash_token_counts(data: &Value) -> (Option<u32>, Option<u32>) {
    // Try top-level `usage` object (e.g. { "usage": { "input_tokens": N, "output_tokens": N } })
    if let Some(usage) = data.get("usage") {
        let tin = usage
            .get("input_tokens")
            .or_else(|| usage.get("prompt_tokens"))
            .and_then(Value::as_u64)
            .map(|n| n as u32);
        let tout = usage
            .get("output_tokens")
            .or_else(|| usage.get("completion_tokens"))
            .and_then(Value::as_u64)
            .map(|n| n as u32);
        if tin.is_some() || tout.is_some() {
            return (tin, tout);
        }
    }
    // Try `message.usage` (some gateway shapes nest it there)
    if let Some(msg_usage) = data.get("message").and_then(|m| m.get("usage")) {
        let tin = msg_usage
            .get("input_tokens")
            .or_else(|| msg_usage.get("prompt_tokens"))
            .and_then(Value::as_u64)
            .map(|n| n as u32);
        let tout = msg_usage
            .get("output_tokens")
            .or_else(|| msg_usage.get("completion_tokens"))
            .and_then(Value::as_u64)
            .map(|n| n as u32);
        if tin.is_some() || tout.is_some() {
            return (tin, tout);
        }
    }
    (None, None)
}

fn request_id_from_payload(data: &Value) -> Option<String> {
    data.get("requestId")
        .and_then(Value::as_str)
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string())
}

pub(crate) async fn post_json(
    endpoint: &str,
    headers: Vec<(String, String)>,
    payload: Value,
) -> Result<(Value, std::collections::HashMap<String, String>), ProviderError> {
    let client = Client::builder()
        .timeout(Duration::from_secs(LLM_TIMEOUT_SECS))
        .build()
        .map_err(|e| ProviderError::Transport(e.to_string()))?;
    let mut req = client.post(endpoint);
    for (k, v) in headers {
        req = req.header(&k, &v);
    }
    let resp = req
        .json(&payload)
        .send()
        .await
        .map_err(|e| ProviderError::Transport(e.to_string()))?;
    let status = resp.status();
    let response_headers: std::collections::HashMap<String, String> = resp
        .headers()
        .iter()
        .filter_map(|(k, v)| {
            v.to_str()
                .ok()
                .map(|s| (k.as_str().to_ascii_lowercase(), s.to_string()))
        })
        .collect();
    if status.as_u16() >= 400 {
        return Err(provider_error(status.as_u16(), resp).await);
    }
    let text = resp
        .text()
        .await
        .map_err(|e| ProviderError::Transport(e.to_string()))?;
    let data: Value = serde_json::from_str(&text).map_err(|_| ProviderError::NonJson)?;
    Ok((data, response_headers))
}

async fn provider_error(status: u16, resp: reqwest::Response) -> ProviderError {
    let rid = resp
        .headers()
        .get("RequestId")
        .or_else(|| resp.headers().get("x-request-id"))
        .and_then(|v| v.to_str().ok())
        .map(|s| s.to_string());
    let text = resp.text().await.unwrap_or_default();
    let mut detail = text.trim().to_string();
    if let Ok(parsed) = serde_json::from_str::<Value>(&detail) {
        if let Some(dv) = parsed
            .get("message")
            .or_else(|| parsed.get("error"))
            .or_else(|| parsed.get("detail"))
        {
            detail = match dv {
                Value::String(s) => s.clone(),
                Value::Object(_) => dv
                    .get("message")
                    .and_then(Value::as_str)
                    .map(str::to_string)
                    .unwrap_or_else(|| dv.to_string()),
                _ => dv.to_string(),
            };
        }
    }
    let rid_suffix = match rid {
        Some(r) => format!(" RequestId: {r}."),
        None => String::new(),
    };
    let detail_suffix = if detail.is_empty() {
        String::new()
    } else {
        format!(": {detail}")
    };
    ProviderError::HttpStatus {
        status,
        detail: detail_suffix,
        rid: rid_suffix,
    }
}
